keps/sig-node/2621-cpu-allocation-llc-affinity/README.md

```diff
@@ -0,0 +1,179 @@
+
+![allocation_motivation](allocation_motivation.png "allocation_motivation")
+
+### Goals
+
+Support uncore-cache affinity in cpu allocation in architecture.
+
+### Future Work
+
+Cross-die may also decrease process performance. We will add die affinity future, and corresponding cpu assignment algorithm implemetation.
+
+## Proposal
+
+In order to make a decision to allocate cpu with uncore-cache affinity, we should be aware of the uncore-cache information in kubelet, current kubelet gets cpu topology with cadvisor, which does not support the related details. So, we add cache id and uncore-cache items to cadvisor(all merged).
+- Add cache id to cadvisor
+In cadvisor PR(https://github.com/google/cadvisor/pull/2847/),  use /sys/devices/system/cpu/cpu*/cache/index3/id to get L3 cache id of current cpu, and store it as cpu topology.
+```go
+type Cache struct {
++	// Id of memory cache
++	Id int `json:"id"`
+	// Size of memory cache in bytes.
+	Size uint64 `json:"size"`
+	// Type of memory cache: data, instruction, or unified.
+	Type string `json:"type"`
+	// Level (distance from cpus) in a multi-level cache hierarchy.
+	Level int `json:"level"`
+}
+```
+- Add uncore cache to cadvisor
+In cadvisor PR(https://github.com/google/cadvisor/pull/2849), add L3 cache not shared among the whole socket(uncore cache) to core info in cpu topology. And we can get core->uncore-cache mappings.
+```go
+type Core struct {
+	Id           int     `json:"core_id"`
+	Threads      []int   `json:"thread_ids"`
+	Caches       []Cache `json:"caches"`
++	UncoreCaches []Cache `json:"uncore_caches"`
+	SocketID     int     `json:"socket_id"`
+}
+```
+
+### User Stories (Optional)
+
+Before change, when kubelet allocates cpus for containers, uncore-cache is not considered, and may get cpus across caches even there're free cpus shared uncore-caches. 
+We make a bench with `stream2` DAXPY, as we can see, cross ccx(cross uncore-cache) gets lower bandwidth.
+
+![stream2_daxpy](stream2_daxpy.png "stream2_daxpy")
+
+And, when workload is memory sensitive, this feature can improve memory bandwidth significantly(20% above).
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
++ Currently no risks was found.
++ Feature is enbled by a gate - a new kube feature with default false, potential risk effects could be limited.
+
+## Design Details
+
+- Feature Gate
+  - Add `CPUManagerUncoreCacheAlign` to kubelet's feature-gates to enable(true)/disable(false) the feature.
+  - Also, more than one l3 cache should exist in a single socket/package.
```

Comment:
It is probably implied but maybe we should explicitly capture the scenario where `CPUManagerUncoreCacheAlign` is enabled and in case only one l3 cache is present, we would obtain the current behaviour.